### PR TITLE
Add implementation for Fastmode+ I2C available on select MCUs

### PR DESCRIPTION
**Pull request by therealprof**
This peripheral is very similar to the one found in newer MCUs e.g.
STM32F0 and was taken from there.

Signed-off-by: Daniel Egger <daniel@eggers-club.de>

### CODE REVIEW

**Comment by therealprof** [comment]
@aurelj @TeXitoi Any opinions on this?

**Comment by aurelj** [comment]
> @aurelj @TeXitoi Any opinions on this?

I haven't looked in details but this seems great overall.

The FMPI2C peripheral definitions in the SVDs are quite messy and inconsistent, so I've tried to clean this up here:
  https://github.com/stm32-rs/stm32-rs/pull/390
This should allow to support FMPI2C on the f410, f412, f413, f423 and f446 very easily.

**Comment by therealprof** [comment]
Wow, that is excellent! I've wanted to implement it for all STM32F4 that support it but ran into those gnarly inconsistencies in naming so it stripped it back to the only ones which compile with the current state.

I've successfully tested it with the STM32F413H Discovery board but only with regular I2C speeds (400kHz).

File: src/i2c.rs [hunk 1]
```diff
@@ -617,6 +671,28 @@ impl<PINS> I2c<I2C3, PINS> {
...
+impl<PINS> FMPI2c<FMPI2C, PINS> {
+    pub fn fmpi2c(i2c: FMPI2C, pins: PINS, speed: KiloHertz, clocks: Clocks) -> Self
+    where
+        PINS: Pins<FMPI2C>,
+    {
+        // NOTE(unsafe) This executes only during initialisation
+        let rcc = unsafe { &(*RCC::ptr()) };
+
+        // Enable clock for FMPI2C
+        rcc.apb1enr.modify(|_, w| w.i2c4en().set_bit());
```
Comment (TeXitoi) [inline src/i2c.rs:684]: Isn't it racy?

File: src/i2c.rs [hunk 2]
```diff
@@ -852,3 +928,240 @@ where
...
+            scldel = 4;
+        } else {
+            presc = 0;
+            scll = cmp::max((((freq >> presc) >> 1) / speed.0) - 1, 255) as u8;
+            sclh = scll - 6;
+            sdadel = 1;
+            scldel = 3;
+        }
+
+        // Enable I2C signal generator, and configure I2C for 400KHz full speed
```
Comment (TeXitoi) [inline src/i2c.rs:967]: Inconsistent comment?

Comment (therealprof) [reply src/i2c.rs:967]: Yes, thanks for noticing. Guess I'll need to fix it in stm32f0xx-hal, too. ;)

File: src/i2c.rs (hunk 1 above)
Comment (therealprof) [reply src/i2c.rs:684]: Yes it is, any suggestions?

Comment (TeXitoi) [reply src/i2c.rs:684]: Critical section, bit banging or &mut Rcc as parameter?

Comment (therealprof) [reply src/i2c.rs:684]: Yeah, the usual suspects. ;)

At least the race is consistent with the already existing other I2C impls. I guess we might want to fix them all in the same way in one swoop?

Comment (TeXitoi) [reply src/i2c.rs:684]: As you wish, there is a lot of racy code in this crate. Maybe we should open issues to track them?

Comment (therealprof) [reply src/i2c.rs:684]: Let's do that.

Comment (therealprof) [reply src/i2c.rs:684]: I don't have the time right now to check all the places which require attention but I've opened #171 to keep track of required fixes.

**Pull request merged**